torch_xla/distributed/spmd/debugging.py

```diff
@@ -0,0 +1,165 @@
+
+
+class ShardingCallbackInfo:
+
+  def __init__(self, callback, module_context):
+    self.callback = callback
+    self.module_context = module_context
+
+
+def visualize_sharding(sharding: str,
+                       use_color: bool = True,
+                       scale: float = 1.,
+                       min_width: int = 9,
+                       max_width: int = 80):
+  """Visualizes a ``Sharding`` using ``rich``.
+  Args:
+    sharding (`str`): sharding of given tensor with SPMD
+    use_color (`bool`): whether use color or not
+    scale (`float`): scale of table visualized in console
+    min_width (`int`): min width used to setup table to visualize
+    max_width (`int`): max width used to setup table to visualize
+  Returns:
+    table to visualize given tensor sharding. This function
+    will also visualize the sharding of the tensor without as return.
+  """
+
+  if not RICH_ENABLED:
+    raise ValueError("`visualize_sharding` requires `rich` to be installed.")
+
+  slices: dict[tuple[int, ...], set[int]] = {}
+  heights: dict[tuple[int, ...], Optional[float]] = {}
+  widths: dict[tuple[int, ...], float] = {}
+
+  if len(sharding) > 0:
+    # sharding is longer than 0
+    # eg: '{devices=[2,2]0,1,2,3}' # 13
+    # eg: '{replicated}'
+    # eg: '{devices=[2,1,2]0,1,2,3 last_tile_dim_replicate}' # 15
+    if sharding == '{replicated}':
+      # eg: '{replicated}'
+      heights = 1
+      widths = 1
+      num_devices = xr.global_runtime_device_count()
+      device_ids = list(range(num_devices))
+      slices.setdefault((0, 0), device_ids)
+    else:
+      # `device_indices_map`: [0, 1, 2, 3]
+      # `sharding_spac`: [2, 2]
+      sharding_spac = sharding[sharding.index('['):sharding.index(']') + 1]
+      if len(sharding) >= 25 and sharding[-24:-1] == 'last_tile_dim_replicate':
+        device_list = sharding[sharding.index(']') + 1:-24]
+        middle = device_list[:-1]
+        device_indices_map = [int(s) for s in middle.split(',')]
+        heights = int(sharding_spac[1])
+        widths = int(sharding_spac[3])
+        last_dim_depth = int(sharding_spac[5])
+        devices_len = len(device_indices_map)
+        len_after_dim_down = devices_len // last_dim_depth
+        for i in range(len_after_dim_down):
+          slices.setdefault(
+              (i // widths, i % widths),
+              device_indices_map[i * last_dim_depth:(i + 1) * last_dim_depth])
+      elif sharding[-1] == "}":
+        # eg: '{devices=[2,2]0,1,2,3}' # 13
+        device_list = sharding[sharding.index(']') + 1:-1]
+        device_indices_map = [int(i) for i in device_list.split(',')]
+        heights = int(sharding_spac[1])
+        widths = int(sharding_spac[3])
+        devices_len = len(device_indices_map)
+        for i in range(devices_len):
+          slices.setdefault((i // widths, i % widths), device_indices_map[i])
+      else:
+        raise ValueError("sharding is not organized as expected")
```

Comment:
Show the actual invalid sharding as part of the err message.

Reply:
thanks, updated